Tests/Unit/NetworkServer/ExclusiveProcessorTests.cs

```diff
@@ -0,0 +1,230 @@
+        }
+
+        [Fact]
+        public async Task ProcessAsync_With_Canceled_Function_Throws_TaskCanceledException()
+        {
+            using var cts = new CancellationTokenSource();
+            cts.Cancel();
+
+            var ex = await Assert.ThrowsAsync<TaskCanceledException>(() =>
+                this.subject.ProcessAsync(1, () => Task.FromCanceled<int>(cts.Token)));
+
+            Assert.Equal(cts.Token, ex.CancellationToken);
+        }
+
+        [Fact]
+        public async Task ProcessAsync_With_Erroneous_Function_Throws_Thrown_Exception()
+        {
+#pragma warning disable CA2201 // Do not raise reserved exception types
+            var ex = new ApplicationException();
+#pragma warning restore CA2201 // Do not raise reserved exception types
+
+            var actual = await Assert.ThrowsAsync<ApplicationException>(() =>
+                this.subject.ProcessAsync(1, () => Task.FromException<int>(ex)));
+
+            Assert.Same(ex, actual);
+        }
+
+        [Fact]
+        public async Task ProcessAsync_Raises_Events()
+        {
+            (object Sender, int Args) submittedEvent = default;
+            (object Sender, int Args) processingEvent = default;
+            (object Sender, (int, ExclusiveProcessor<int>.IProcessingOutcome) Args) processedEvent = default;
+
+            this.subject.Submitted += (sender, args) => submittedEvent = (sender, args);
+            this.subject.Processing += (sender, args) => processingEvent = (sender, args);
+            this.subject.Processed += (sender, args) => processedEvent = (sender, args);
+
+            const int id = 1;
+            var outcome = await this.subject.ProcessAsync(id, () => Task.FromResult(42));
+
+            Assert.Same(this.subject, submittedEvent.Sender);
+            Assert.Equal(id, submittedEvent.Args);
+
+            Assert.Same(this.subject, processingEvent.Sender);
+            Assert.Equal(id, processingEvent.Args);
+
+            Assert.Same(this.subject, processedEvent.Sender);
+            var (_, (processedId, processedOutcome)) = processedEvent;
+            Assert.Equal(id, processedId);
+            Assert.True(processedOutcome.Task.IsCompletedSuccessfully);
+            Assert.Equal(outcome.SubmissionTime, processedOutcome.SubmissionTime);
+            Assert.Equal(outcome.WaitDuration, processedOutcome.WaitDuration);
+            Assert.Equal(outcome.RunDuration, processedOutcome.RunDuration);
+        }
+
+        [Theory]
+        [InlineData(10)]
+        public async Task ProcessAsync_Is_Fifo_By_Default(int processorCount)
+        {
+            var processorIds = Enumerable.Range(1, processorCount).ToArray();
+            var (submittedList, processedList) = await TestProcessing(this.subject, processorIds);
+
+            Assert.Equal(processedList, submittedList);
+            Assert.Equal(processedList, processorIds);
+        }
+
+        private sealed class LifoScheduler<T> : ExclusiveProcessor<T>.IScheduler
```

Comment:
[nit] Do you think it would be adding a test to assert at which point the interrupt event is going to be called? I think it might be worth it as "documentation", I had to debug through the code to understand when it's expected.

Reply:
Those events are purely for diagnostics and incidentally also help with tests. They are not critical to the overall operation. The other events are not tested either so I think adding for `Interrupted` would make it look somewhat exceptional. I would even opt for removing the even entirely.